lib/spack/spack/test/cmd/env.py

```diff
@@ -3119,6 +3120,46 @@ def test_environment_depfile_makefile(depfile_flags, expected_installs, tmpdir,
     assert len(specs_that_make_would_install) == len(expected_installs)
 
 
+def test_environment_depfile_spec_format_special_chars(tmpdir, mock_packages, monkeypatch):
+    env("create", "test")
+    makefile = str(tmpdir.join("Makefile"))
+    with ev.read("test"):
+        add("dttop")
+        concretize()
+
+    orig_format = spack.spec.Spec.format
+
+    def _contrived_spec_format(obj, format_str):
+        if obj.name == "dtlink1":
+            return "dtlink1-githash=version-spackhash"
+        else:
+            return orig_format(obj, format_str)
+
+    monkeypatch.setattr(spack.spec.Spec, "format", _contrived_spec_format)
+
+    # Disable jobserver so we can do a dry run.
```

Comment:
You're not using dry run 🤔 tbh, dry run is the best way to validate the makefile, and this is an integration test already, so it might even be better than manually validating the generated output

Reply:
Dry run could be useful to do in addition to this, but I strongly feel that it is not enough: the problem I ran into involved a makefile which you could use (no syntax errors, and running `make` proceeds), but was entirely wrong (e.g. would perform installations of packages in the wrong order); I don't think a dry run could catch this. It could be useful for detecting other problems with generating the makefile though.

Reply:
Can you show an example? I don't entirely see how it would not result in make error 🤔

Reply:
I would need to check but I think the makefile generated by this test should successfully execute a dry run without the changes outside of the unit-tests (I didn't check that as part of this PR because (a) the special characters were the source of the problem in the larger example and (b) I removed those as part of the PR).

Reply:
> instead of verifying the makefile behavior I think it is more-appropriate to verify that no unexpected special characters are introduced.

Sure, but imo an actual unit test would be best then: concrete specs in, adjacency list out; verify nodes, their names, and edges. That requires a bit of refactoring.

> This seems to me to be nearly as invasive

It's not. The minimal change is `2 insertions(+), 1 deletion(-)`, and could've been a oneliner if spack wasn't so insistent on needing a `git` prop in the mock package:

```diff
diff --git a/lib/spack/spack/test/cmd/env.py b/lib/spack/spack/test/cmd/env.py
index 8fb16e8322..662b87d74c 100644
--- a/lib/spack/spack/test/cmd/env.py
+++ b/lib/spack/spack/test/cmd/env.py
@@ -3091,7 +3091,7 @@ def test_environment_depfile_makefile(depfile_flags, expected_installs, tmpdir,
     make = Executable("make")
     makefile = str(tmpdir.join("Makefile"))
     with ev.read("test"):
-        add("dttop")
+        add("dttop ^dtlink1@aaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaa=1.2.3")
         concretize()
 
     # Disable jobserver so we can do a dry run.
diff --git a/var/spack/repos/builtin.mock/packages/dtlink1/package.py b/var/spack/repos/builtin.mock/packages/dtlink1/package.py
index 836ca59e69..a8c1d100b5 100644
--- a/var/spack/repos/builtin.mock/packages/dtlink1/package.py
+++ b/var/spack/repos/builtin.mock/packages/dtlink1/package.py
@@ -11,6 +11,7 @@ class Dtlink1(Package):
 
     homepage = "http://www.example.com"
     url = "http://www.example.com/dtlink1-1.0.tar.gz"
+    git = "http://www.example.com/dtlink1.git"
 
     version("1.0", "0123456789abcdef0123456789abcdef")
```

Reply:
> The minimal change is 2 insertions(+), 1 deletion(-), and could've been a oneliner if spack wasn't so insistent on needing a git prop in the mock package:

I want to be clear that I strongly recommend against doing anything like what this diff suggests: ultimately it does not matter what properties inside of specs give rise to special characters, only that they could be generated as part of using Spec functionality to generate a makefile target.

There is a better way to do what I did (and I'm happy to try to make that happen), but it is not this particular diff.

>> instead of verifying the makefile behavior I think it is more-appropriate to verify that no unexpected special characters are introduced.

> Sure, but imo an actual unit test would be best then: concrete specs in, adjacency list out; verify nodes, their names, and edges. That requires a bit of refactoring.

My initial interpretation of what you said is to examine something like `MakeTargetVisitor.adjacency_list`, but really any test that examines entities before the final makefile will be fragile: you have to know which properties to check; if the makefile generation introduces an extra reference to the spec anywhere then that would be another potential source of unwanted special characters.

I could test `_fmt_spec_make_target` directly (I think this would be closer to what you consider a unit test vs. an integration test), although if that isn't called in some spot where it should be, the makefile will again end up with special characters.

I could refactor the `env_depfile` function so that it only ever interacts with a new class like `MakefileSpec`, which could be mocked by the tests to ensure that users never call `str` on it, and only get special characters if they call an explicit function like `MakefileSpec.unsafe_format()`. Overall something like.

```
def env_depfile():
    roots = [s for _, s in env.concretized_specs()]
    all_specs = list(traverse(roots)) # No visitor here: need to convert type
    roots = [MakefileSpec(x) for x in roots]
    all_specs = [MakefileSpec(x) for x in all_specs]
    # At this point we have more control over how the makefile could pull details from the spec
    generate_makefile(roots, all_specs)
```

does that seem better?

Reply:
If you both have an integration test for git versions, then this

> if the makefile generation introduces an extra reference to the spec anywhere then that would be another potential source of unwanted special characters.

is not a concern in the unit test.

If I would write it I would just make a small class (the model) that computes the properties that are passed to `template.render` (the view) and then unit test the model properties.

Reply:
At the same time, I also don't care too much at this point 😆 as long as the bug fix gets in

Reply:
> If I would write it I would just make a small class (the model) that has properties that are passed to template.render (the view)

I think you're talking about taking the dictionary passed to `template.render` and turning it into an object:

```
    rendered = template.render(
        {
            "all_target": get_target("all"),
            "env_target": get_target("env"),
```

Validating such an object would be complex: the targets in general can have several different special characters, and while I don't think `=` is among them (which makes it easy to look for that specifically as a problem), there are other special characters that can be generated as part of `Spec.format` which *are* used in targets (e.g. `%`). Therefore, this object would need to differentiate between the spec and non-spec portions of targets, i.e. have functions like:

```
class Depfile(object):
    def add_spec_target(self, preamble: str, spec: Spec):
        # the preamble can contain "%" or anything we want, and isn't validated
        ...

    def add_target(self, target: str):
        # Hopefully no one called str() on a Spec to get the target
        ....
```

I think my suggestion from https://github.com/spack/spack/pull/36642#discussion_r1164510858 would simplify the matter:

> I could refactor the env_depfile function so that it only ever interacts with a new class like MakefileSpec (rather than Spec), 

The `MakefileSpecs` (which are not full specs and so can be mocked without monkeypatching) can be fed into the functions more or less the same as before to generate the template input dictionary; I could split the logic to then stop before rendering the template (to me, creating the template does not make this an integration test, although from discussion here I gather you think it does).

Reply:
> Validating such an object would be complex

I'm not sure I follow 🤔 I've implemented

> If I would write it I would just make a small class (the model) that has properties that are passed to template.render (the view)

in #36911, and it's not complex to do `data = MakefileModel(...).to_dict()` and test correctness of `data`. You could mock MakefileModel's `Spec` -> `str` function to produce something distinct and verify that it pops up in `data` for entries that are targets, etc.